Snakefile

```diff
@@ -20,24 +22,77 @@ validate(config, schema="schemas/config_schema.yaml")
 # CADD environment variable
 import os
 
+###################################
+## RULE SPECFIC THREADING LIMITS ##
+###################################
+system_memory = psutil.virtual_memory().total / (1024 ** 3)
+gpu_memory = 0
+try:
+    lines = subprocess.check_output("nvidia-smi --query-gpu=memory.total --format=csv,noheader,nounits 2>/dev/null", shell=True).decode('utf-8').splitlines()
+    # sum over gpu(s)
+    gpu_memory = int(sum([int(x) for x in lines])/1024)
+except Exception as e:
+    pass
+    
+# esm takes ~16 Gb of system memory per tread & 4 Gb of GPU ram (if available). 
+if gpu_memory > 0:
+    config['esm_slots'] = int(min((0.9*system_memory)/16, 0.95*gpu_memory/4))
+else:
+    config['esm_slots'] = int(0.9*system_memory / 16)
+if config['esm_slots'] < 1:
+    config['esm_slots'] = 1
+# then assign other resources
+config['esm_load'] = 100 if config['esm_slots'] < 1 else int(100/config['esm_slots']) 
+config['esm_threads'] = int(workflow.cores /  config['esm_slots'])
+config['vep_load'] = 100 if system_memory < 4 else int(100/int(0.9*system_memory / 4 ))  # up to 4Gb/ram
+config['regseq_load'] = 100 if system_memory < 2 else int(100/int(0.9*system_memory / 2 ))   # up to 2Gb of ram
+config['mms_load'] = 100 if system_memory < 16 else int(100/int(0.9*system_memory / 16 ))   # up to 16Gb/ram
+config['mms_threads'] = int(workflow.cores / (100/config['mms_load']))
+config['anno_load'] = 1 # disk IO intensive
+config['impute_load'] = 1 
+config['prescore_load'] = 1 
+config['score_load'] = 1
+
+print("Threading Overview",flush=True)
+print("##################",flush=True)
+print("Assigned cores: {}".format(workflow.cores),flush=True)
+print("Available system memory: {}GB".format(int(system_memory)),flush=True)
+if gpu_memory > 0:
+   print("Total GPU memory: {}GB".format(gpu_memory),flush=True)
+else:
+   print("No gpu found",flush=True)
+print("Task Parallelization: ",flush=True)
+print("  PreScore : {}x".format(min(workflow.cores,int(100/config['prescore_load']))))
+print("  VEP : {}x".format(min(workflow.cores,int(100/config['vep_load']))))
+print("  ESM : {}x with {} threads each (memory/gpu constraints)".format(min(workflow.cores,config['esm_slots']),config['esm_threads']))
+print("  RegSeq : {}x".format(min(workflow.cores,int(100/config['regseq_load']))))
+print("  MMsplice : {}x with {} threads each (memory constraints)".format(min(workflow.cores,int(100/config['mms_load'])),config['mms_threads']))
+print("  Annotate : {}x".format(min(workflow.cores,int(100/config['anno_load']))))
+print("  Impute : {}x".format(min(workflow.cores,int(100/config['impute_load']))))
+          
+
+
+
+## allowed scattering 
+scattergather:
+    split=workflow.cores,
 
 envvars:
     "CADD",
 
 
-# wildcard_constraints:
-#     file=".*(?<!\\.vcf)$"
+#wildcard_constraints:
+#    basefile="[^/]+"
```

Comment:
Why?

Reply:
that's some leftover i missed. It's commented in both version, so no impact. but it can be reverted